File: typical90/typical90_z.py
# https://atcoder.jp/contests/typical90/tasks/typical90_z

import sys
# from numba import njit
from collections import Counter
sys.setrecursionlimit(10 ** 7)

N = int(input())
G = [[] for _ in range(N)]
for _ in range(N-1):
    a, b = map(int, input().split())
    G[a-1].append(b-1)
    G[b-1].append(a-1)

col = [-1] * N

def dfs(pos, color):
    col[pos] = color
    for nxt in G[pos]:
        if col[nxt]==-1:
            dfs(nxt, 1 - color)
    return

dfs(0, 0)
print(*col, file=sys.stderr)
cnt = Counter(col)
if cnt[0]>=cnt[1]:
    ans = [i+1 for i in range(N) if col[i]==0]
else:
    ans = [i+1 for i in range(N) if col[i]==1]
print(*ans[:N//2])


# TLE
# import sys
# input = sys.stdin.buffer.readline
# sys.setrecursionlimit(10 ** 7)

# MAX_N = 100005
# col = [-1] * MAX_N
# to = [[] for _ in range(MAX_N)]

# def dfs(pos, color):
#     col[pos] = color
#     for nxt in to[pos]:
#         if col[nxt]==-1:
#             dfs(nxt, 1-color)
#     return

# N = int(input())
# for i in range(N-1):
#     a, b = map(int, input().split())
#     to[a-1].append(b-1)
#     to[b-1].append(a-1)

# dfs(0, 0)
# zeros, ones = 0, 0
# for i in range(N):
#     if col[i]==0: zeros += 1
#     else:         ones += 1
# if zeros>=ones:
#     c = 0
# else:
#     c = 1
# cnt = 0
# ans = []
# for i in range(N):
#     if col[i]==c:
#         ans.append(i+1)
#         cnt += 1
#     if cnt*2==N:
#         break
# print(*ans)

File: typical90/typical90_z.cpp
// https://atcoder.jp/contests/typical90/tasks/typical90_z
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
// #include <atcoder/all>
// using namespace atcoder;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
#define maxs(x,y) (x = max(x,y))
#define mins(x,y) (x = min(x,y))
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

const int MAX_N = 100005;
vector<int> col;
vector<vector<int>> to;

void dfs(int pos, int color) {
    // cout << "dfs" << endl;
    col[pos] = color;
    for (int nxt: to[pos]) {
        if (col[nxt]==-1) dfs(nxt, 1-color);
    }
    return;
}

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
	int n;
	cin >> n;
    col = vector<int>(n, -1);
    to.resize(n);
    // cout << "check0" << endl;
    rep(i, n-1) {
        int a, b;
        cin >> a >> b;
        a--; b--;
        to[a].push_back(b);
        to[b].push_back(a);
    }
    // cout << "check1" << endl;
    // for (int nxt: to[0]) cout << nxt << endl;
    dfs(0, 0);
    // rep(i,n) cout << col[i] << endl;

    // cout << "check2" << endl;
    int zeros = 0, ones = 0;
    rep(i, n) {
        if (col[i]==0) zeros++;
        else           ones++;
    }
    int c = 1;
    if (zeros*2>=n) c = 0;
    int cnt = 0;
    rep(i, n) {
        if (col[i]==c) {
            cout << i+1 << ' ';
            cnt++;
        }
        if (cnt*2>=n) break;
    }
	return 0;
}
